src/Compiler/Checking/Expressions/CheckExpressions.fs

```diff
@@ -6378,9 +6376,8 @@ and TcExprSequentialOrImplicitYield (cenv: cenv) overallTy env tpenv (sp, synExp
             | Expr.DebugPoint(_,e) -> e
             | _ -> expr1
 
-        env.eCachedImplicitYieldExpressions.Add(synExpr1.Range, (synExpr1, expr1Ty, cachedExpr))
-        try TcExpr cenv overallTy env tpenv otherExpr
-        finally env.eCachedImplicitYieldExpressions.Remove synExpr1.Range
+        (getImplicitYieldExpressionsCache cenv).AddOrUpdate(synExpr1, (expr1Ty, cachedExpr))
```

Comment:
The semantics w.r.t. to error handling is different now, aren't they?

Also it feels like the previous version had a race condition (`FindAll` and later `Add`), possibly the `Multi` map was chosen to workaround it by storing a list and not a single value?

Reply:
Previous version had some concurrency problems, because Vlad at some point added `ConcurrentDictionary` as `HashMultiMap` backing store to fix them.

Semantics are different. Originally this just always removed the value in `finally`. I have to revisit this, I remember the change made sense to me but I forgot why, oh lol. Now I also notice this was attached to `env`, not `cenv`.

This is not really tested in the test suite but there is a benchmark.  I'll run it later to see if this still works.